GitUI/Avatars/CustomAvatarProvider.UriTemplateResolver.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Drawing;
+using System.Threading.Tasks;
+using System.Web;
+using JetBrains.Annotations;
+
+namespace GitUI.Avatars
+{
+    public sealed partial class CustomAvatarProvider
+    {
+        /// <summary>
+        /// An avatar provider that takes <see cref="UriTemplateData"/> as input to provide <see cref="Uri"/>s.
+        /// </summary>
+        /// <remarks>
+        /// The template resolver claims to be an <see cref="IAvatarProvider"/> but that's not really true.
+        /// It only implements the interface to allow the parent class <see cref="CustomAvatarProvider"/> to
+        /// store providers of two different types without using <see cref="object"/> as base type.
+        /// A cleaner alternative would be for <see cref="CustomAvatarProvider"/> to use a discriminated union,
+        /// but sadly C# doesn't offer them at the moment and external dependency just for that case
+        /// would probably be overkill.
+        /// </remarks>
+        private class UriTemplateResolver : IAvatarProvider
+        {
+            private readonly Func<UriTemplateData, string> _templateResolver;
+
+            public UriTemplateResolver(string uriTemplate)
+            {
+                // create a formatter, so we only have to parse the template once
+                // and reuse the formatter for a lot of inputs / user.
+                _templateResolver = BuildResolver(uriTemplate);
+            }
+
+            public Task<Image> GetAvatarAsync([NotNull] string email, string name, int imageSize)
+            {
+                // see remarks for details why we do this.
+                throw new InvalidOperationException($"UriTemplateResolvers can't be used as a regular IAvatarProvider, use '{nameof(ResolveTemplate)}'.");
+            }
```

Comment:
Should not be IAvatarProvider. We need to change implementation.

Reply:
I commented why it is the way it is. The only alternatives are:
- Using `object` as base type
- Use a discriminated union (which I would have to implement only for that use-case, or we would need a third party library that supports that)

Reply:
Not return from FromTemplateSegment.